src/preprocessing.ml

```diff
@@ -477,9 +477,70 @@ let get_compat_ppx_exe sctx ~name ~kind =
     in
     ppx_exe sctx ~key ~dir_kind:Jbuild
 
-let get_ppx_driver sctx ~loc ~scope ~dir_kind pps =
+let args_of_cookies cookies =
+  List.map ~f:(fun (name, value) ->
+    ["--cookie"; sprintf "%s=%S" name value]
+  ) cookies |> List.flatten
+
+let library_name_cookie : Lib_kind.Ppx_args.Cookie.t =
+  {
+    name = "library_name";
+    value = String_with_vars.make_var Loc.none "library_name"
+  }
+let get_cookies ~loc ~expander ~lib_name libs =
+  let expander =
+    match lib_name with
+    | None -> expander
+    | Some name ->
+      let bindings =
+        Pform.Map.singleton "library_name"
+          (Values [String (Lib_name.Local.to_string name)])
+      in
+      Expander.add_bindings expander ~bindings
+  in
+  try
+    Ok (libs
+      |> List.map ~f:
+        (fun t ->
+          match Lib.kind t with
+          | Normal -> []
+          | Ppx_rewriter {cookies}
+          | Ppx_deriver {cookies} ->
+            List.map ~f:(fun {Lib_kind.Ppx_args.Cookie.name; value} ->
+              (name, (Expander.expand_str expander value, Lib.name t)))
+              (if lib_name = None then cookies else library_name_cookie :: cookies)
```

Comment:
It seems that we are adding the library name cookie once for each library. I suppose that wasn't intended? In particular, that's probably why the library name cookie disappears in some of the integration tests.